scripts/income_limits/ami.py

```diff
@@ -0,0 +1,192 @@
+    .drop('county_name', axis=1)
+)
+
+zcta_countysub_crosswalk = (
+    util.clean_colnames(pd.read_csv(
+        util.DATA_FPATH / 'raw' / 'geocorr2022_zcta_to_countysub.csv',
+        encoding="ISO-8859-1",
+        low_memory=False,
+        skiprows=1,
+        dtype={col: str for col in string_cols}))
+    .rename({'zip_census_tabulation_area': 'zcta'}, axis=1)
+    .drop('county_name', axis=1)
+)
+
+# remove non-zctas
+zcta_countysub_crosswalk = zcta_countysub_crosswalk[
+    zcta_countysub_crosswalk.zip_code_name != '[not in a ZCTA]']
+
+# clean up geoids
+zcta_countysub_crosswalk['countysub_geoid'] = zcta_countysub_crosswalk[['county_code',
+                                                                        'county_subdivision_2020']].agg(''.join, axis=1)
+tract_countysub_crosswalk['countysub_geoid'] = tract_countysub_crosswalk[['county_code',
+                                                                          'county_subdivision_2020']].agg(''.join, axis=1)
+tract_countysub_crosswalk['tract_geoid'] = tract_countysub_crosswalk.apply(
+    lambda x: x.county_code + f"{x.tract:.2f}".replace('.', '').zfill(6), axis=1)
+
+# -- 3. Map countysub AMIs to target geographies (ZCTA/tract) and aggregate over target -- #
+
+# Join amis to zctas and to tracts, where new england states will match on countysub geoids
+# and non-new england states will match on county geoids
+# The following to not have a match in the crosswalk, and thus get dropped
+# * 15 unpopulated countysubs in New England
+# * Sullivan MO which is an edge case of a county sub outside New England
+# TODO: add data checks for future refreshes?
+
+# bool mask indicating if row is a county (True) or a county subdivision (False)
+is_county_msk = ami_by_countysub.town_name == ''
+
+# join county and countysub amis to zcta
+ami_county_zcta_non_new_england = (
+    ami_by_countysub[is_county_msk]
+    .merge(
+        zcta_countysub_crosswalk.drop('countysub_geoid', axis=1),
+        left_on='county_geoid',
+        right_on='county_code')
+)
+ami_countysub_zcta_new_england = (
+    ami_by_countysub[~is_county_msk]
+    .merge(
+        zcta_countysub_crosswalk,
+        on='countysub_geoid')
+)
+ami_countysub_zcta = pd.concat(
+    [ami_county_zcta_non_new_england, ami_countysub_zcta_new_england])
+
+# join county and countysub amis to tract
+ami_county_tract_non_new_england = (
+    ami_by_countysub[is_county_msk]
+    .merge(
+        tract_countysub_crosswalk.drop('countysub_geoid', axis=1),
+        left_on='county_geoid',
+        right_on='county_code')
+)
+ami_countysub_tract_new_england = (
+    ami_by_countysub[~is_county_msk]
+    .merge(
+        tract_countysub_crosswalk,
+        on='countysub_geoid')
+)
+ami_countysub_tract = pd.concat(
+    [ami_county_tract_non_new_england, ami_countysub_tract_new_england])
+# %%
+# Aggregate over countysubs to get AMI at target geography using strategy of minimizng type II error
+ami_by_zcta = util.aggregate_over_origin(
```

Comment:
Blocking: I attempted to use the other implementation - the weighted average over origin geographies, and it threw the following error: 

`TypeError: aggregate_over_origin.<locals>.weighted_average() got an unexpected keyword argument 'include_groups'` 

Seems you might not need the `include_groups` arg in the weighted average function?

Additionally, have you characterized the difference between the result using both methodologies? I attempted to here but wasn't able to resolve the error, but might be good to examine if adjacent counties or countysubs have wildly different AMI thresholds. I feel like it's unlikely, but would be good to know generally.

Reply:
Gotcha, my python version was out of date